File: tool/render_contents_icons.py
#!/usr/bin/env python3
"""Render Sotong Contents Promo brand icons from SVG sources."""

from __future__ import annotations

import hashlib
import shutil
from pathlib import Path

import cairosvg
from PIL import Image

ROOT = Path(__file__).resolve().parents[1]
SRC_ICON = ROOT / "web" / "icons" / "sotong_contents_icon.svg"
SRC_MASKABLE = ROOT / "web" / "icons" / "sotong_contents_icon_maskable.svg"
BRAND_COPY = ROOT / "assets" / "branding" / "sotong_contents_icon.svg"

OUTPUTS = [
    (SRC_ICON, ROOT / "web" / "favicon.png", 48),
    (SRC_ICON, ROOT / "web" / "icons" / "Icon-192.png", 192),
    (SRC_ICON, ROOT / "web" / "icons" / "Icon-512.png", 512),
    (SRC_MASKABLE, ROOT / "web" / "icons" / "Icon-maskable-192.png", 192),
    (SRC_MASKABLE, ROOT / "web" / "icons" / "Icon-maskable-512.png", 512),
]


def render_png(svg_path: Path, out_path: Path, size: int) -> None:
    png_bytes = cairosvg.svg2png(
        url=str(svg_path),
        output_width=size,
        output_height=size,
    )
    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_bytes(png_bytes)

    with Image.open(out_path) as img:
        if img.size != (size, size):
            raise RuntimeError(f"{out_path} size is {img.size}, expected {(size, size)}")
        if img.mode not in ("RGB", "RGBA", "P"):
            raise RuntimeError(f"{out_path} unexpected mode {img.mode}")
        if out_path.stat().st_size <= 0:
            raise RuntimeError(f"{out_path} is empty")


def sha256(path: Path) -> str:
    return hashlib.sha256(path.read_bytes()).hexdigest()


def main() -> None:
    if not SRC_ICON.exists():
        raise FileNotFoundError(SRC_ICON)
    if not SRC_MASKABLE.exists():
        raise FileNotFoundError(SRC_MASKABLE)

    BRAND_COPY.parent.mkdir(parents=True, exist_ok=True)
    shutil.copyfile(SRC_ICON, BRAND_COPY)

    print("Rendering icons...")
    for svg_path, out_path, size in OUTPUTS:
        render_png(svg_path, out_path, size)
        with Image.open(out_path) as img:
            print(
                f"OK {out_path.relative_to(ROOT)} "
                f"{img.size[0]}x{img.size[1]} mode={img.mode} "
                f"bytes={out_path.stat().st_size} sha256={sha256(out_path)[:12]}"
            )

    print("Done.")


if __name__ == "__main__":
    main()

File: tool/render_contents_icons.js
#!/usr/bin/env node
/**
 * Render Sotong Contents Promo brand icons from SVG sources.
 * Requires: npm install --prefix tool @resvg/resvg-js
 */
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { Resvg } = require('@resvg/resvg-js');

const ROOT = path.resolve(__dirname, '..');
const SRC_ICON = path.join(ROOT, 'web', 'icons', 'sotong_contents_icon.svg');
const SRC_MASKABLE = path.join(ROOT, 'web', 'icons', 'sotong_contents_icon_maskable.svg');
const BRAND_COPY = path.join(ROOT, 'assets', 'branding', 'sotong_contents_icon.svg');

const OUTPUTS = [
  { svg: SRC_ICON, out: path.join(ROOT, 'web', 'favicon.png'), size: 48 },
  { svg: SRC_ICON, out: path.join(ROOT, 'web', 'icons', 'Icon-192.png'), size: 192 },
  { svg: SRC_ICON, out: path.join(ROOT, 'web', 'icons', 'Icon-512.png'), size: 512 },
  { svg: SRC_MASKABLE, out: path.join(ROOT, 'web', 'icons', 'Icon-maskable-192.png'), size: 192 },
  { svg: SRC_MASKABLE, out: path.join(ROOT, 'web', 'icons', 'Icon-maskable-512.png'), size: 512 },
];

function sha256(filePath) {
  return crypto.createHash('sha256').update(fs.readFileSync(filePath)).digest('hex');
}

function pngSize(buf) {
  if (buf[0] !== 0x89 || buf[1] !== 0x50) {
    throw new Error('Not a PNG');
  }
  const width = buf.readUInt32BE(16);
  const height = buf.readUInt32BE(20);
  return { width, height };
}

function renderOne({ svg, out, size }) {
  const svgData = fs.readFileSync(svg);
  const resvg = new Resvg(svgData, {
    fitTo: { mode: 'width', value: size },
  });
  const pngData = resvg.render().asPng();
  fs.mkdirSync(path.dirname(out), { recursive: true });
  fs.writeFileSync(out, pngData);

  const { width, height } = pngSize(pngData);
  if (width !== size || height !== size) {
    throw new Error(`${out} size is ${width}x${height}, expected ${size}x${size}`);
  }
  if (pngData.length <= 0) {
    throw new Error(`${out} is empty`);
  }

  const rel = path.relative(ROOT, out);
  console.log(
    `OK ${rel} ${width}x${height} bytes=${pngData.length} sha256=${sha256(out).slice(0, 12)}`
  );
}

function main() {
  if (!fs.existsSync(SRC_ICON)) throw new Error(`Missing ${SRC_ICON}`);
  if (!fs.existsSync(SRC_MASKABLE)) throw new Error(`Missing ${SRC_MASKABLE}`);

  fs.mkdirSync(path.dirname(BRAND_COPY), { recursive: true });
  fs.copyFileSync(SRC_ICON, BRAND_COPY);
  console.log('Copied branding SVG');

  console.log('Rendering icons...');
  for (const job of OUTPUTS) {
    renderOne(job);
  }
  console.log('Done.');
}

main();
